### PR TITLE
Mkw issue 76

### LINKED ISSUE #76: Fix hardcoding of dimension for art state spaces
- [x] hc_art_total, hc_art_need_init, and hc_art_init all have a dimension of 4 which needs to be moved to a specific dim that is named. 
- [x] Also need to see if all of these hc_art variables are needed and provide firm definitions

### PR DESCRIPTION
**Pull request by mwalte10**
Closes #76 

- Removed an unused state space object
- Added some more state space variables to reduce hard coding in model_input and model_output files

### CODE REVIEW

File: inst/include/child_model_simulation.hpp [hunk 1]
```diff
@@ -169,6 +169,26 @@ void calc_hiv_negative_pop(int time_step,
...
+                           State<ModelVariant, real_type> &state_next,
+                           IntermediateData<ModelVariant, real_type> &intermediate) {
+  const auto demog = pars.base.demography;
+  const auto cpars = pars.children.children;
+  static_assert(ModelVariant::run_child_model,
+                "adjust_hiv_births can only be called for model variants where run_child_model is true");
+  constexpr auto ss = StateSpace<ModelVariant>().base;
+
+if(cpars.abortion(time_step,1) == 1){
+  state_next.children.hiv_births -= state_next.children.hiv_births * cpars.abortion(time_step,0);
```
Comment (M-Kusumgar) [inline inst/include/child_model_simulation.hpp:185]: if you wanna be slick with it (completely optional change for fun)
```suggestion
  state_next.children.hiv_births *= 1 - cpars.abortion(time_step,0);
```

**Review by M-Kusumgar — APPROVED** [review]
Looks good to me!

File: inst/include/child_model_simulation.hpp (hunk 1 above)
Comment (mwalte10) [reply inst/include/child_model_simulation.hpp:185]: To be honest I will forget what this does in the future.

File: inst/cpp_generation/model_input.csv [hunk 2]
```diff
@@ -24,8 +24,8 @@ paed_cd4_dist,Children.hc1_cd4_dist,real_type,ModelVariant::run_child_model,,FAL
 paed_cd4_transition,Children.hc_cd4_transition,real_type,ModelVariant::run_child_model,,FALSE,2,children.hc2DS,children.hc1DS,,
 paed_cd4_mort,Children.hc1_cd4_mort,real_type,ModelVariant::run_child_model,,FALSE,3,children.hc1DS,children.hcTT,children.hc1AG,
 adol_cd4_mort,Children.hc2_cd4_mort,real_type,ModelVariant::run_child_model,,FALSE,3,children.hc2DS,children.hcTT,children.hc2AG,
-paed_cd4_prog,Children.hc1_cd4_prog,real_type,ModelVariant::run_child_model,,FALSE,1,children.hc1DS,,,
-adol_cd4_prog,Children.hc2_cd4_prog,real_type,ModelVariant::run_child_model,,FALSE,1,children.hc2DS,,,
+paed_cd4_prog,Children.hc1_cd4_prog,real_type,ModelVariant::run_child_model,,FALSE,3,children.hc1DS,2,base.NS,
+adol_cd4_prog,Children.hc2_cd4_prog,real_type,ModelVariant::run_child_model,,FALSE,3,children.hc2DS,1,base.NS,
```
Comment (r-ash) [inline inst/cpp_generation/model_input.csv:28]: These are now hardcoded, what are the 2 and 1 in the 2nd dimension? Do we need a 1 would that not be the same as `children.hc2DS x base.NS`?

File: scripts/create_test_data.R [hunk 3]
```diff
@@ -15,20 +15,20 @@ source('./scripts/read_spectrum.R')
...
+#
+# # Used as reference data (Run from leapfrog/master)
+# lmod <- leapfrog::leapfrogR(demp, proj)
+# saveRDS(lmod, testthat::test_path("testdata/leapfrog_fit.rds"))
+#
+# lmod <- leapfrog::leapfrogR(demp, proj, hiv_strat = "coarse")
+# saveRDS(lmod, testthat::test_path("testdata/leapfrog_fit_coarse.rds"))
+#
+# mod <- leapfrog::leapfrogR(demp, proj, hiv_steps_per_year = 0L)
+# saveRDS(lmod, testthat::test_path("testdata/fit_demography.rds"))
```
Comment (r-ash) [inline scripts/create_test_data.R:31]: Can we uncomment this?

File: tests/testthat/testdata/bwa_aim-no-special-elig-numpmtct.PJNZ
Comment (r-ash) [inline tests/testthat/testdata/bwa_aim-no-special-elig-numpmtct.PJNZ:1]: Is this file used for anything?

File: tests/testthat/testdata/bwa_aim-no-special-elig-numpmtct_pop1.xlsx
Comment (r-ash) [inline tests/testthat/testdata/bwa_aim-no-special-elig-numpmtct_pop1.xlsx:1]: Maybe I am missing it but this file also looks like it is not used?

**Review by r-ash — CHANGES_REQUESTED** [review]

File: inst/cpp_generation/model_input.csv (hunk 2 above)
Comment (mwalte10) [reply inst/cpp_generation/model_input.csv]: Both of these represent the number of coarse age groups. For paed_cd4_prog, this is 0-2 year olds and 3-4 year olds. I've added variables for these now.

File: scripts/create_test_data.R (hunk 3 above)
Comment (mwalte10) [reply scripts/create_test_data.R]: Yep

File: tests/testthat/testdata/bwa_aim-no-special-elig-numpmtct.PJNZ
Comment (mwalte10) [reply tests/testthat/testdata/bwa_aim-no-special-elig-numpmtct.PJNZ:1]: This is the file used for the paed model testing.

File: tests/testthat/testdata/bwa_aim-no-special-elig-numpmtct_pop1.xlsx
Comment (mwalte10) [reply tests/testthat/testdata/bwa_aim-no-special-elig-numpmtct_pop1.xlsx:1]: Also used for paed model testing.

File: inst/include/child_model_simulation.hpp [hunk 4]
```diff
@@ -390,12 +424,13 @@ void run_calculate_perinatal_transmission_rate(int time_step,
...
-    intermediate.children.age_weighted_infections +=  demog.age_specific_fertility_rate(a, time_step) / intermediate.children.asfr_sum  * state_curr.base.p_infections(a + 15,1) ; //newly infected 15-49 women, weighted for ASFR
+    intermediate.children.age_weighted_infections +=  demog.age_specific_fertility_rate(a, time_step) / intermediate.children.asfr_sum  * state_next.base.p_infections(a + 15,1) ; //newly infected 15-49 women, weighted for ASFR
   }//end
 
   if (intermediate.children.age_weighted_hivneg > 0.0) {
     intermediate.children.incidence_rate_wlhiv = intermediate.children.age_weighted_infections / intermediate.children.age_weighted_hivneg;
-    intermediate.children.perinatal_transmission_from_incidence = intermediate.children.incidence_rate_wlhiv * (9/12) * (intermediate.children.births_sum - intermediate.children.need_PMTCT) * cpars.vertical_transmission_rate(7,0);
+    //0.75 is 9/12, gestational period
+    intermediate.children.perinatal_transmission_from_incidence = intermediate.children.incidence_rate_wlhiv * 0.75 * (intermediate.children.births_sum - intermediate.children.need_PMTCT) * cpars.vertical_transmission_rate(7,0);
```
Comment (jeffeaton) [inline inst/include/child_model_simulation.hpp:433]: What is the `(7,0)` here?  Wonder if it is useful to have a descriptive macro for that

File: scripts/read_spectrum.R [hunk 5]
```diff
@@ -11,9 +11,18 @@ prepare_hc_leapfrog_projp <- function(pjnz, params, pop_1){
   v$paed_cd4_dist <- dp_read_paed_cd4_dist(pjnz) / 100
 
   prog = dp_read_paed_cd4_prog(pjnz)
-  paed_cd4_prog <- array(c(prog[1,1:6],0), dim = 7, dimnames = list(cd4 = c('30plus', '26-30', '21-25', '16-20', '11-15', '5-10', '<5')))
+  paed_cd4_prog <- array(0, dim = c(7,2,2), dimnames = list(cd4 = c('30plus', '26-30', '21-25', '16-20', '11-15', '5-10', '<5'),
```
Comment (jeffeaton) [inline scripts/read_spectrum.R:14]: Are these CD4 percent categories?  Maybe label the name as `list(cd4percent = ...` if so.

Also are 26-30, 21-25, etc. accurate? Vs. 25-29, 20-24, etc.

File: scripts/read_spectrum.R [hunk 6]
```diff
@@ -11,9 +11,18 @@ prepare_hc_leapfrog_projp <- function(pjnz, params, pop_1){
...
+                                                            age = c('0-2', '3-4'),
+                                                            sex = c('male', 'female')))
+  paed_cd4_prog[,1,1] <- c(prog[1,1:6], 0)
+  paed_cd4_prog[,1,2] <- c(prog[2,1:6], 0)
 
-  adol_cd4_prog <- array(c(prog[1,14:18],0), dim = 6, dimnames = list(cd4 = c('>1000', '750-999', '500-749', '350-499', '200-349', 'lte200')))
```
Comment (jeffeaton) [inline scripts/read_spectrum.R:16]: calling 5-14 year olds 'adolescents' is a bit of a stretch. But I guess it's a simple clear distinction.

File: scripts/read_spectrum.R [hunk 7]
```diff
@@ -294,6 +304,7 @@ prepare_hc_leapfrog_projp <- function(pjnz, params, pop_1){
   v$prop_lt200 <- colSums(wlhiv_cd4[5:7,]) / colSums(wlhiv_cd4)
 
   v$hc_age_coarse <- rep(c(1,2,3), each = 5)
+  v$hc_age_coarse_cd4 <- c(0,0,0,1,1,rep(2,10))
```
Comment (jeffeaton) [inline scripts/read_spectrum.R:307]: Might code this as `rep(0:2, times = c(3, 2, 10))`. But inconsequential.

**Review by jeffeaton — APPROVED** [review]
Thanks, this looks good. Some inconsequential cosmetic comments for at your discretion.

File: scripts/read_spectrum.R (hunk 5 above)
Comment (mwalte10) [reply scripts/read_spectrum.R:14]: changed name to cd4percent, 

This stratification is what is used in Spectrum

File: inst/include/child_model_simulation.hpp (hunk 4 above)
Comment (mwalte10) [reply inst/include/child_model_simulation.hpp:433]: Added a comment, represents the index with transmission rate for maternal seroconversion

File: scripts/read_spectrum.R (hunk 7 above)
Comment (mwalte10) [reply scripts/read_spectrum.R:307]: Sure

**Pull request merged**